File: infer/tests/codetoanalyze/java/quandary/Hil.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.quandary;

import java.util.ArrayList;
import java.util.List;

/**
 * Repro of HIL bindings issue: In the exception node, the invariant that a bound variable cannot
 * appear in the RHS of another binding is broken.
 */
class Hil {
  public static void foo() {
    List<Integer> Y = new ArrayList<>();
    int dummy = 0;
    for (Character x : "X".toCharArray()) {
      dummy = 1;
    }
    for (Integer y : Y) {
      dummy = 2;
    }
  }
}

File: infer/tests/codetoanalyze/kotlin/racerd/Hil.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

class Hil(public val o: Object) {
  fun crashInfer(nums: List<Long>) {
    val foldedNums = nums.fold(1) { _, _ -> 2 }

    val checked = checkNotNull(o) { "Boom" }

    val v = foldedNums.toLong()
  }
}
